src/main/java/org/truffleruby/core/hash/DeleteLastNode.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.core.hash;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import org.truffleruby.language.RubyContextNode;
+
+import com.oracle.truffle.api.dsl.ImportStatic;
+import com.oracle.truffle.api.dsl.Specialization;
+
+@ImportStatic(HashGuards.class)
+public abstract class DeleteLastNode extends RubyContextNode {
+    /* For BucketHash, behaves almost identical to DeleteNode. Otherwise, deletes the most recently added key, because
+     * we are using the hash like a stack and guarantee that the last we added is the one we'll want to delete. When
+     * assertions are enabled checks that the deleted key is the one expected by the user. Does not handle blocks. A
+     * helper for detect_recursion. */
+
+    @Child private LookupEntryNode lookupEntryNode = new LookupEntryNode();
+
+    public static DeleteLastNode create() {
+        return DeleteLastNodeGen.create();
+    }
+
+    public abstract Object executeDeleteLast(RubyHash hash, Object key);
+
+    @Specialization(guards = "isNullHash(hash)")
+    protected Object deleteNull(RubyHash hash, Object key) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        throw new UnsupportedOperationException("Cannot delete the last node of an empty hash");
+    }
+
+    @Specialization(guards = "isPackedHash(hash)")
+    protected Object deletePackedArray(RubyHash hash, Object key) {
+        assert HashOperations.verifyStore(getContext(), hash);
+
+        final Object[] store = (Object[]) hash.store;
+        final int size = hash.size;
+
+        int n = size - 1;
+
+        // removable
+        final Object otherKey = PackedArrayStrategy.getKey(store, n);
+        assert key == otherKey;
+
+        final Object value = PackedArrayStrategy.getValue(store, n);
+        PackedArrayStrategy.removeEntry(getLanguage(), store, n);
+        hash.size -= 1;
+        assert HashOperations.verifyStore(getContext(), hash);
+        return value;
+    }
+
+    @Specialization(guards = "isBucketHash(hash)")
+    protected Object delete(RubyHash hash, Object key) {
+        assert HashOperations.verifyStore(getContext(), hash);
+
+        final HashLookupResult hashLookupResult = lookupEntryNode.lookup(hash, key);
```

Comment:
If we wanted to optimize more here, we could probably avoid the lookup entirely, and just use `hash.lastInSequence`.
The entry also contains the `hashed` value so, we can find the previous bucket entry that way.

Reply:
It only seems to save us a `int hashed = hashNode.execute(key, compareByIdentity);`, so I personally feel like the opt isn't worth the repeated code, but I can definitely do the opt if I'm mistaken or the opt feels worth it to others.

Reply:
I think potentially `hashNode.execute()` is not so fast, depending on the type of the key, and it would also be possible to skip `eql?` calls as well, since we already know the `Entry`, which could be a big saving.
Regarding duplication, we could add a `static LookupEntryNode.lookup(RubyHash hash, Entry entry)`, so the similar code would stay next to each other. It's also not a lot of code.

Reply:
You're right! the save was bigger than I thought. :D